### ISSUE TITLE
[Security] Object properties stored on UserInterface may be inconsistent or inaccurate

### ISSUE THREAD

**Issue by arderyp** [symfony/symfony#27459]
**Symfony version(s) affected**: 2.8.x - 3.4.x (maybe 4 as well, haven't tested that)

**Description**  

I have a custom Guard user called `GuardUser` that implements `UserInterface`. I have my own local user entity called `User`.  I pass my `User` entity into my `GuardUser` object during instantiation in order to mirror some of the `User` properties on the `GuardUser` object.  Simple `User` properties like strings and integers map just fine.  But, when I try to mirror object properties on `GuardUser`, particularly deep Doctrine Collection objects, the data consistency can start to fall apart.

**How to reproduce**  

When I instantiate the `GuardUser` in my custom `UserProvider` , I pass the corresponding `User` entity into a `hydrateFromUser` method where I take properties from the `User` entity and apply them to the `GuardUser` object:

```
use App\Package\Entity\User;

class GuardUser implements UserInterface
{
    private $address;
    private $childCollection;
    private $user;

    ....
    public function hydrateFromUser(User $user)
    {
        $this->address = $user->getAddress(); // string
        $this->childCollection = $user->getChildCollection();  // ManyToMany Doctrine collection
        $this->user = $user; // User

        var_dump(get_class($this->user));  // outputs "App\Package\Entity\User"
    }

    public function getUserInfo()
    {
        var_dump(gettype($this->address));  // this data is correct

        var_dump(gettype($this->childCollection));  // the objects in this collection are inconsistent in some contexts

        var_dump(gettype($this->user));  // now this is a string, not an object

        ...
    }
    ....
}
```

Later in the session, on some specific page, I call the `getUserInfo` method on my `GuardUser` instance for the current authenticated used, and now the local `user` property on that object is a string instead of an instance of `App\Package\Entity\User`.

The issue with `childCollection` is rather complex, and confusing.  I am happy to share details if someone thinks they can help, but I'm kind of hoping someone might point out something obvious that I am doing wrong that might resolve all of this weirdness I am experiencing.  If need be, I can try to "bug reproducer" route, but I fear that I won't be able to get that working, as the issue related to the `Collections` properties is quite complex and thorny, and depends on various relational data mappings.

**Possible Solution**  

no clue

**Additional context**

**Comment by iltar** [comment]
Don't put complexity in your security user. If you need the additional data, query it based on your user. If anything, don't even store (non value) objects in your user, it will just cause problems with unserialization and such.

**Comment by arderyp** [comment]
thanks @iltar, that seems to be exactly what I am experiencing.  Hopefully I can restructure things to use a better approach.

**Comment by arderyp** [comment]
**UPDATE**: I ended up implementing a new approach in my code, following @iltar's advice to keep my `UserInterface` implementation object as simple as possible, with no object properties, to facilitate easy serialization.

So, I am not sure what the devs/maintainers want to do about this ticket.  I am not sure if there is actually a bug, or if its just a given that serialization may be problematic with big/complex object properties.  Feel free to close if you want.

All this being said, should the documentation ([this](https://symfony.com/doc/current/security/custom_provider.html), [this](https://symfony.com/doc/3.4/security/entity_provider.html#understanding-serialize-and-how-a-user-is-saved-in-the-session), etc) include a note/best practice about keeping custom `UserInterface` implementations as simple as possible?

I wonder if there might also be security concerns with serializing sensitive data on the session (another argument for keeping the custom `UserInterface` implementation as small/simple as possible).  Also worth considering... will programing languages move away from serialization interpreters ([article](https://adtmag.com/articles/2018/05/30/java-serialization.aspx))?

**Comment by iltar** [comment]
> All this being said, should the documentation include a note/best practice about keeping custom `UserInterface` implementations as simple as possible?

In my opinion, the docs should be as simple as possible. That means that entities should not be used in the examples to show the security user or form data for the primary documentation. They can always be added to show the coupling in an additional example. There are so many questions about entities in forms and the security user coming by, as both introduce a very complex coupling with entities.

I wrote these two blog posts for that particular reason:
 - https://stovepipe.systems/post/avoiding-entities-in-forms
 - https://stovepipe.systems/post/decoupling-your-security-user

**Comment by arderyp** [comment]
Great blog posts @iltar, particularly the second one, which applies directly to the subject of this ticket.  I wish I'd found them sooner.  Thankfully, I came to the same solution as you did in your blog post (more or less) in building and refactoring my authentication system.  

So yeah, your blog post seems to confirm my suspicion that the issue I reported here originally isn't really a symfony bug, per se.  Its simply fair to assume that serialization problems will pop up when complex object properties are involved**.  I think this could be expressed in a single sentence in the official documentation to prevent people from pursuing problematic implementations (and very thorny ones to debug, at that!), but I also understand your desire to keep the documentation short and sweet.

 ** @iltar and @chalasr, if you agree with this point, feel free to close this issue.
